### PR TITLE
fix: add scheduler for retry

### PR DESCRIPTION
**Pull request by minz1027**
Scheduler is needed for Retry plugin.
This PR will add scheduler and end it before the process ends.

Docs for node-resque 4.X, it's on v4 branch: https://github.com/taskrabbit/node-resque/tree/v4

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -100,13 +108,34 @@ multiWorker.on('internalError', error =>
...
+scheduler.on('error', error =>
+    winston.info(`scheduler error >> ${error}`));
+scheduler.on('working_timestamp', timestamp =>
+    winston.info(`scheduler working timestamp ${timestamp}`));
+scheduler.on('transferred_job', (timestamp, job) =>
+    winston.info(`scheduler enquing job timestamp  >>  ${JSON.stringify(job)}`));
+
 multiWorker.start();
+scheduler.connect(() => {
+    scheduler.start();
```
Comment (minz1027) [inline index.js]: Don't know why the `start` syntax is different for `multiworker` and `scheduler`. But this is what they put in their doc.

File: index.js [hunk 2]
```diff
@@ -45,22 +45,28 @@ function updateBuildStatus(updateConfig, callback) {
 }
 
 /**
- * Shutdown workers and then exit the process
+ * Shutdown both worker and scheduler and then exit the process
  * @method shutDownWorker
```
Comment (tkyi) [inline index.js]: shutDownAll*

File: index.js [hunk 3]
```diff
@@ -45,22 +45,28 @@ function updateBuildStatus(updateConfig, callback) {
...
  * @param  {Object}       worker    worker to be ended
+ * @param  {Object}       schduler  schduler to be ended
```
Comment (d2lam) [inline index.js]: misspelled lol

Comment (tkyi) [inline index.js]: typo: scheduler*

File: index.js [hunk 4]
```diff
@@ -100,13 +108,34 @@ multiWorker.on('internalError', error =>
...
+scheduler.on('poll', () =>
+    winston.info('scheduler polling'));
+scheduler.on('master', state =>
+    winston.info(`scheduler became master ${state}`));
+scheduler.on('error', error =>
+    winston.info(`scheduler error >> ${error}`));
+scheduler.on('working_timestamp', timestamp =>
+    winston.info(`scheduler working timestamp ${timestamp}`));
+scheduler.on('transferred_job', (timestamp, job) =>
+    winston.info(`scheduler enquing job timestamp  >>  ${JSON.stringify(job)}`));
```
Comment (tkyi) [inline index.js]: enqueuing*

File: index.js [hunk 5]
```diff
@@ -45,22 +45,28 @@ function updateBuildStatus(updateConfig, callback) {
...
+            winston.error(`failed to end the worker: ${error}`);
         }
 
-        process.exit(0);
+        scheduler.end((err) => {
+            if (err) {
+                winston.error(`failed to end the scheduler: ${err}`);
+                process.exit(128);
+            }
+            process.exit(0);
```
Comment (d2lam) [inline index.js]: Doesn't this mean if worker fails to end, but as long as scheduler ends successfully, it still `exit(0)`?

File: test/index.test.js [hunk 6]
```diff
@@ -129,28 +141,43 @@ describe('Index Test', () => {
             });
         });
 
-        it('logs error and exit with non-zero code when it fails to end worker', (done) => {
+        it('logs error and then end schduler when it fails to end worker', (done) => {
```
Comment (tkyi) [inline test/index.test.js]: scheduler*

File: test/index.test.js [hunk 7]
```diff
@@ -194,6 +221,36 @@ describe('Index Test', () => {
...
+            assert.calledWith(winstonMock.info,
+                `scheduler error >> ${error}`);
+
+            testScheduler.emit('working_timestamp', timestamp);
+            assert.calledWith(winstonMock.info,
+                `scheduler working timestamp ${timestamp}`);
+
+            testScheduler.emit('transferred_job', timestamp, job);
+            assert.calledWith(winstonMock.info,
+                `scheduler enquing job timestamp  >>  ${JSON.stringify(job)}`);
```
Comment (tkyi) [inline test/index.test.js]: enqueuing*

File: test/index.test.js [hunk 8]
```diff
@@ -213,16 +270,26 @@ describe('Index Test', () => {
...
             process.once('SIGTERM', () => {
-                assert.calledOnce(shutDownWorkerMock);
+                assert.calledOnce(shutDownAllMock);
                 done();
             });
             process.kill(process.pid, 'SIGTERM');
         });
     });
+
+    describe('schduler', () => {
```
Comment (tkyi) [inline test/index.test.js]: scheduler*

File: index.js [hunk 9]
```diff
@@ -100,13 +108,34 @@ multiWorker.on('internalError', error =>
...
+
 multiWorker.start();
+scheduler.connect(() => {
+    scheduler.start();
+});
 
 // Shut down workers before exit the process
-process.on('SIGTERM', () => supportFunction.shutDownWorker(multiWorker));
+process.on('SIGTERM', () => {
+    supportFunction.shutDownAll(multiWorker, scheduler);
```
Comment (d2lam) [inline index.js]: Any reason why we need to switch to this syntax?

File: index.js [hunk 10]
```diff
@@ -100,13 +108,34 @@ multiWorker.on('internalError', error =>
...
+    winston.info('scheduler started'));
+scheduler.on('end', () =>
+    winston.info('scheduler ended'));
+scheduler.on('poll', () =>
+    winston.info('scheduler polling'));
+scheduler.on('master', state =>
+    winston.info(`scheduler became master ${state}`));
+scheduler.on('error', error =>
+    winston.info(`scheduler error >> ${error}`));
+scheduler.on('working_timestamp', timestamp =>
```
Comment (d2lam) [inline index.js]: Can you give us the link to the current doc for this version? Current README shows as `workingTimestamp`...

**Comment by FenrirUnbound** [comment]
It seems like the scheduler can be created independent of the worker. Do we want the scheduler to be tied together with the worker?

**Comment by minz1027** [comment]
@FenrirUnbound yes, it can be on it's own. I think we should definitely separate them when we have multiple replicas of workers. For now, I don't see a big benefit to separate them.

File: index.js (hunk 10 above)
Comment (minz1027) [reply index.js]: There is no link in current README, their npm doc is still old, I put it as reference in the PR description. https://www.npmjs.com/package/node-resque

File: index.js (hunk 5 above)
Comment (minz1027) [reply index.js]: yes, but we are not doing anything about the exit code. So this actually doesn't matter

File: index.js (hunk 9 above)
Comment (minz1027) [reply index.js]: no.. previously I implemented in some other way, forgot to change it back

File: index.js (hunk 10 above)
Comment (minz1027) [reply index.js]: Or you can clone this repo and do npm install and check inside the node_module. That's how I figure it out.. But couldn't find an easy link on their repo

**Pull request merged**